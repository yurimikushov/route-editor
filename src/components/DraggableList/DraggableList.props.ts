import { ReactNode } from 'react'

type DraggableListProps<T> = {
  className?: string
  draggableClassName?: string
  list: Array<Item<T>>
  renderItem: (item: Item<T>, isDraggable: boolean) => ReactNode
  onUpdate: (list: Array<Item<T>>) => void
}

type Item<T> = T & { id: string }

export default DraggableListProps
