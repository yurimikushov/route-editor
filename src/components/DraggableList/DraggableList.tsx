import {
  DragDropContext,
  Droppable,
  Draggable,
  DropResult,
} from 'react-beautiful-dnd'
import cn from 'classnames'
import map from 'lodash/map'
import isNil from 'lodash/isNil'
import reorder from 'lib/reorder'
import DraggableListProps from './DraggableList.props'

const DraggableList = <T,>({
  className,
  draggableClassName,
  list,
  renderItem,
  onUpdate,
}: DraggableListProps<T>) => {
  const handleDragEnd = (result: DropResult) => {
    if (isNil(result.destination)) {
      return
    }

    onUpdate(reorder(list, result.source.index, result.destination.index))
  }

  return (
    <DragDropContext onDragEnd={handleDragEnd}>
      <Droppable droppableId='draggable-list'>
        {(provided, snapshot) => (
          <ul
            ref={provided.innerRef}
            {...provided.droppableProps}
            className={cn(className, {
              [draggableClassName ?? '']: snapshot.isDraggingOver,
            })}
          >
            {map(list, (item, index) => (
              <Draggable key={item.id} draggableId={item.id} index={index}>
                {(provided) => (
                  <li
                    ref={provided.innerRef}
                    {...provided.draggableProps}
                    {...provided.dragHandleProps}
                    tabIndex={-1}
                  >
                    {renderItem(item)}
                  </li>
                )}
              </Draggable>
            ))}
            {provided.placeholder}
          </ul>
        )}
      </Droppable>
    </DragDropContext>
  )
}

export default DraggableList
