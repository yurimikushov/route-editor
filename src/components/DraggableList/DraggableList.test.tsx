import renderer from 'react-test-renderer'
import noop from 'lodash/noop'
import DraggableList from './DraggableList'

it(`should render draggable points when route contains more than 1 element`, () => {
  const list = [
    {
      id: '1',
      title: 'Item #1',
    },
    {
      id: '2',
      title: 'Item #2',
    },
  ]

  const tree = renderer
    .create(
      <DraggableList
        className='some-draggable-list'
        draggableClassName='dragging-modifier'
        list={list}
        renderItem={(item, isDragging) => (
          <div>
            <div>{item.title}</div>
            {isDragging && <div>Dragger</div>}
          </div>
        )}
        onUpdate={noop}
      />
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it(`should render non-draggable point when route contains 1 element`, () => {
  const list = [
    {
      id: '1',
      title: 'Item #1',
    },
  ]

  const tree = renderer
    .create(
      <DraggableList
        className='some-draggable-list'
        draggableClassName='dragging-modifier'
        list={list}
        renderItem={(item, isDragging) => (
          <div>
            <div>{item.title}</div>
            {isDragging && <div>Dragger</div>}
          </div>
        )}
        onUpdate={noop}
      />
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it(`should render empty container when list is empty`, () => {
  const list = [] as Array<{ id: string; title: string }>

  const tree = renderer
    .create(
      <DraggableList
        className='some-draggable-list'
        draggableClassName='dragging-modifier'
        list={list}
        renderItem={(item, isDragging) => (
          <div>
            <div>{item.title}</div>
            {isDragging && <div>Dragger</div>}
          </div>
        )}
        onUpdate={noop}
      />
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
