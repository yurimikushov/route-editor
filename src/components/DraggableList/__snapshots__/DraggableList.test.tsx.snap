// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render draggable points when route contains more than 1 element 1`] = `
<ul
  className="some-draggable-list"
  data-rbd-droppable-context-id="0"
  data-rbd-droppable-id="draggable-list"
>
  <li
    aria-describedby="rbd-hidden-text-0-hidden-text-0"
    data-rbd-drag-handle-context-id="0"
    data-rbd-drag-handle-draggable-id="1"
    data-rbd-draggable-context-id="0"
    data-rbd-draggable-id="1"
    draggable={false}
    onDragStart={[Function]}
    onTransitionEnd={null}
    role="button"
    style={
      Object {
        "transform": null,
        "transition": null,
      }
    }
    tabIndex={-1}
  >
    <div>
      <div>
        Item #1
      </div>
      <div>
        Dragger
      </div>
    </div>
  </li>
  <li
    aria-describedby="rbd-hidden-text-0-hidden-text-0"
    data-rbd-drag-handle-context-id="0"
    data-rbd-drag-handle-draggable-id="2"
    data-rbd-draggable-context-id="0"
    data-rbd-draggable-id="2"
    draggable={false}
    onDragStart={[Function]}
    onTransitionEnd={null}
    role="button"
    style={
      Object {
        "transform": null,
        "transition": null,
      }
    }
    tabIndex={-1}
  >
    <div>
      <div>
        Item #2
      </div>
      <div>
        Dragger
      </div>
    </div>
  </li>
</ul>
`;

exports[`should render empty container when list is empty 1`] = `
<ul
  className="some-draggable-list"
  data-rbd-droppable-context-id="2"
  data-rbd-droppable-id="draggable-list"
/>
`;

exports[`should render non-draggable point when route contains 1 element 1`] = `
<ul
  className="some-draggable-list"
  data-rbd-droppable-context-id="1"
  data-rbd-droppable-id="draggable-list"
>
  <li
    data-rbd-draggable-context-id="1"
    data-rbd-draggable-id="1"
    onTransitionEnd={null}
    style={
      Object {
        "transform": null,
        "transition": null,
      }
    }
    tabIndex={-1}
  >
    <div>
      <div>
        Item #1
      </div>
    </div>
  </li>
</ul>
`;
